src/pymor/reductors/era.py

```diff
@@ -0,0 +1,238 @@
+        U, sv, V = spla.svd(to_matrix(NumpyHankelOperator(h)), full_matrices=False)
+        return sv, U.T, V
+
+    def output_projector(self, l1):
+        """Construct the left/output projector :math:`W_1`."""
+        assert isinstance(l1, int) and l1 <= self.data.shape[1]
+        self.logger.info(f'Constructing output projector ({l1} tangential directions) ...')
+        return self._s1_W1()[1][:, :l1]
+
+    def input_projector(self, l2):
+        """Construct the right/input projector :math:`W_2`."""
+        assert isinstance(l2, int) and l2 <= self.data.shape[2]
+        self.logger.info(f'Constructing input projector ({l2} tangential directions) ...')
+        return self._s2_W2()[1][:, :l2]
+
+    def error_bounds(self, l1=None, l2=None):
+        r"""Compute the error bounds for all possible reduction orders.
+
+        Without tangential projection of the Markov parameters, the error bounds are defined as
+
+        .. math::
+            \sum_{i=1}^{2s-1}\lVert C_rA_r^{i-1}B_r-h_i\rVert_F^2\leq\sigma_{r+1}(\mathcal{H})
+            \sqrt{r+p+m},
+
+        where :math:`(A_r,B_r,C_r)` is the reduced realization of order :math:`r`,
+        :math:`h_i\in\mathbb{R}^{p\times m}` is the :math:`i`-th Markov parameter
+        and :math:`\sigma_{r+1}(\mathcal{H})` is the first neglected singular value of the
+        Hankel matrix of Markov parameters.
+
+        With tangential projection, the bound is given by
+
+        .. math::
+            \sum_{i=1}^{2s-1}\lVert C_rA_r^{i-1}B_r-h_i\rVert_F^2\leq
+            4\left(\sum_{i=l_1+1}^p\sigma_i^2(\Theta_L)+\sum_{i=l_2+1}^m\sigma_i^2(\Theta_R)\right)
+            +2\sigma_{r+1}(\mathcal{H})\sqrt{r+l_1+l_2},
+
+        where :math:`\Theta_L,\,\Theta_R` is the matrix of horizontally or vertically stacked Markov
+        parameters, respectively. See :cite:`KG16` (Thm. 3.4) for details.
+        """
+        n, p, m = self.data.shape
+        s = n if self.force_stability else (n + 1) // 2
+
+        sv = self._sv_U_V(l1, l2)[0]
+
+        a = p * s if l2 is None and p * s < m else (l2 or m)
+        b = m * s if l1 is None and m * s < p else (l1 or p)
+        err = (np.sqrt(np.arange(len(sv)) + a + b) * sv)[1:]
+
+        err = 2 * err if l1 or l2 else err
+        if l1:
+            s1 = self._s1_W1()[0]
+            err += 4 * np.linalg.norm(s1[l1:])**2
+        if l2:
+            s2 = self._s2_W2()[0]
+            err += 4 * np.linalg.norm(s2[l2:])**2
+
+        return err
+
+    def reduce(self, r=None, tol=None, l1=None, l2=None):
+        """Construct a minimal realization.
+
+        Parameters
+        ----------
+        r
+            Order of the reduced model if `tol` is `None`, maximum order if `tol` is specified.
+        tol
+            Tolerance for the error bound if `r` is `None`.
+        l1
+            Number of left (output) directions for tangential projection.
+        l2
+            Number of right (input) directions for tangential projection.
```

Comment:
Could we come up with better names for `l1` and `l2` (lowercase L is confusing to read)? How about `num_left` and `num_right`?